ASoC: SOF: Intel: Suport for IMR boot and multicore with IPC4

ujfalusi opened the pull request:
Hi,

Add the needed code to enable IMR boot when IPC4 is used (it is always available with IPC4 on supported HDA platforms).
The firmware prepares for the IMR boot on response to a SET_DX message which turns the primary core off (this is the time where IPC4 firmware knows that the DSP is going to be turned off, so it is also doing a context save if applicable).

Since the core_get/core_put ops is implemented for the IMR boot we will also gain support for multicore handling.

ranj063 commented on include/sound/sof/ipc4/header.h:
```diff
@@ -385,6 +385,14 @@ struct sof_ipc4_fw_version {
 
+/* Payload data for SOF_IPC4_MOD_SET_DX */
+struct sof_ipc4_dx_state_info {
+	/* core(s) to aply the change */
```
typo aply

ranj063 commented on sound/soc/sof/intel/cnl.c:
```diff
@@ -361,6 +364,10 @@ int sof_cnl_ops_init(struct snd_sof_dev *sdev)
+
+		/* dsp core get/put */
+		sof_cnl_ops.core_get = hda_dsp_ipc4_core_get;
+		sof_cnl_ops.core_put = hda_dsp_ipc4_core_put;
```
@ujfalusi if we ever tested SOF IPC4 on CNL, this might break because we know multi-core implementation for CNL/APL in SOF is not complete.

ujfalusi replied:
@ranj063, is that's the case for IPC4 as well or only IPC3?
This is the case with APL also, no? I have verified that APL with the closed source firmware we have working IMR support but I don't have CNL device around to test.

ranj063 replied:
@ujfalusi in the case of SOF firmware, it will be the case for both IPC3 and IPC4. I dont know if we want to invest time in fixing multi-core implementation for the older platforms

plbossart replied:
let's enable multi-core for TGL+, we don't have a reason to spend time on multicore for earlier platforms at the moment, we already have a full plate with IPC4 in general.

ujfalusi replied:
@plbossart, @ranj063, this PR takes a simplistic approach to tackle IMR boot by implementing and using the generic core_put() callback to let the firmware know that we are need to prepare for IMR.
It is a side effect that for multicore support we also need the core_put(), but if I'm not mistaken the actual multicore support depends on the loaded topology. Even if we have core_put() implemented for the platform it is not going to matter if the topology we are loading does not use multicore.
I did not wanted to implement another IPC4 wrapper for sending the SET_DX.
If we want to keep IMR, but leave out core_put() then I don't see other way around this. I don't think we want to disable IMR just because multicore is not stable on the given platform with SOF?

ranj063 replied:
@ujfalusi but the point is when we start testing SOF+IPC4, these will fail. Why not stick to IMR boot from TGL onwards because of this limitation

ujfalusi replied:
@ranj063, what will fail with SOF.IPC4? the IPC message for non primary core or the host operation?

If we load a topology without multicore support and have the core_put() implemented, does that going to fail? DO you know if we have some platform flag to say that multicore is not supported or is it only that we don't support the core_put on some platforms?

We can try to remove support for IMR boot when core_put() is not implemented for IPC4, but would not it be a regression? APL with the closed source firmware works fine with IMR for example and if SOF.IPC4 would support APL, I would expect to work there as well. Afaik IPC3 on APL supports IMR boot as well, no?

Let me put this back to the drawing board a bit to figure out a clean way to separate the IMR and multicore (core_put) support.

ranj063 replied:
@ujfalusi if we don't enable multicore in tplg and do a core put for core 0 only.we.might be ok but something to try with ipc4 and SOF. Let me try it out today

ranj063 commented on sound/soc/sof/ipc4.c:
```diff
@@ -597,11 +597,21 @@ static void sof_ipc4_rx_msg(struct snd_sof_dev *sdev)
 	}
 }
 
+static int sof_ipc4_ctx_save(struct snd_sof_dev *sdev)
```
why do you say "as an example" in the commit message? It is exactly how it is done isnt it?

ujfalusi replied:
Yes, it is how it is done, dropped the 'as an example' part.

plbossart commented:
SOFCI TEST

RanderWang commented on sound/soc/sof/intel/hda-dsp.c:
```diff
@@ -1044,3 +1045,64 @@ int hda_dsp_core_get(struct snd_sof_dev *sdev, int core)
+	msg.data_size = sizeof(dx_state);
+
+	ret = sof_ipc_tx_message(sdev->ipc, &msg, msg.data_size, NULL, 0);
+
```
if ret < 0, do we need to return it without doing the following operations ?

ujfalusi replied:
We ignore the error for the primary core and proceed to turn it off. For the non primary cores we return with the error code, but don't do anything on the host side.
We need to send the SET_DX message for all cores, but the host only manages the primary one.

ujfalusi commented:
Changes since v1:
- typo fixed in first patch
- commit message updated for the second patch.

greg-intel commented:
SOFCI TEST

greg-intel commented:
SOFCI TEST

ujfalusi commented:
Changes since v2:
- complete rewrite ;)
- add new IPC pm_ops for setting core state (set_core_state)
 - implementation of set_core_state for IPC3 and IPC4 (IPC message crafting and sending)
- change the intel/hda-dsp.c and intel/tgl.c to use the new IPC pm_ops.set_core_state
 - this takes care of multicore support
- the pm_ops.ctx_save for IPC4 is calling the internal `sof_ipc4_set_core_state(sdev, SOF_DSP_PRIMARY_CORE, false);` directly

Tested on TGL (upx-i11) and APL (up2)

ranj063 commented on sound/soc/sof/sof-priv.h:
```diff
@@ -378,10 +378,12 @@ struct sof_ipc_fw_tracing_ops {
  * struct sof_ipc_pm_ops - IPC-specific PM ops
  * @ctx_save:		Function pointer for context save
  * @ctx_restore:	Function pointer for context restore
+ * @set_core_state:	Function pointer for turning on/off a DSP core
```
@ujfalusi this isnt really for state management but purely for state setting. The management really happens in the core_get/put ops

ujfalusi replied:
@ranj063, you are right. This is atm just a simple change this core to be on.off.
I'll reword the texts accordingly. Thanks for spotting it!

ujfalusi replied:
@ranj063, @plbossart, looking at the code and I think with a different follow-up PR we can drop the tgl specific `tgl_dsp_core_get()` and `tgl_dsp_core_put()` and use the generic HDA `hda_dsp_core_get()`, it is doing exactly the same thing, just written in a different way. We can implement `hda_dsp_core_put()` following the tgl implementation.
I don't see the reason why we would need to have duplicated code.

ranj063 reviewed (dismissed):
Just a nit-pick in the commit message for the first patch but other than that this version is perfect!

ujfalusi commented:
Changes since v3:
- Commit message revised as advised by @ranj063 for the first commit.

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -378,10 +378,12 @@ struct sof_ipc_fw_tracing_ops {
 struct sof_ipc_pm_ops {
 	int (*ctx_save)(struct snd_sof_dev *sdev);
 	int (*ctx_restore)(struct snd_sof_dev *sdev);
+	int (*set_core_state)(struct snd_sof_dev *sdev, int core_idx, bool on);
```
we should mark those as optional/mandatory, this can be done in a follow-up patch @ujfalusi 

in one of the last patches you set ctx_save but not ctx_restore, this is worthy of comments for the rest of us.

ujfalusi replied:
I'll prepare a fixup series to to mark these optional in the struct doc.

plbossart commented on sound/soc/sof/ipc4.c:
```diff
@@ -617,7 +617,13 @@ static int sof_ipc4_set_core_state(struct snd_sof_dev *sdev, int core_idx, bool
+}
+
 static const struct sof_ipc_pm_ops ipc4_pm_ops = {
+	.ctx_save = sof_ipc4_ctx_save,
```
it's unclear why we don't have a restore, but that comment can be added later in a fixup patch.

ujfalusi replied:
OK, I will add a fixup to add a comment.

Sorry, for some reason I have missed your comments and failed to address them in time.

plbossart reviewed (approved):
@ujfalusi see suggestions below to improve comments. I think we should merge and test, comments can be added later.
